File: app/routes/matches.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import or_
from sqlalchemy.orm import Session

from app.database.database import get_db
from app.models.job import Job
from app.models.resume import ResumeProfile
from app.models.user import User
from app.schemas.match import JobMatchResponse
from app.services.matching_service import calculate_match

router = APIRouter(
    prefix="/matches",
    tags=["Matched Jobs"]
)


@router.get("/user/{user_id}", response_model=list[JobMatchResponse])
def get_matched_jobs(user_id: int, db: Session = Depends(get_db)):
    user = db.query(User).filter(User.id == user_id).first()

    if not user:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="User not found"
        )

    resume_profile = db.query(ResumeProfile).filter(
        ResumeProfile.user_id == user_id
    ).first()

    if not resume_profile:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Resume profile not found"
        )

    jobs_query = db.query(Job)

    if user.preferred_location:
        jobs_query = jobs_query.filter(
            or_(
                Job.location.ilike(f"%{user.preferred_location}%"),
                Job.work_mode.ilike("%remote%"),
                Job.work_mode.ilike("%hybrid%")
            )
        )

    matched_jobs = []

    for job in jobs_query.all():
        match_data = calculate_match(
            resume_profile.skills,
            job.required_skills
        )

        if match_data["match_percentage"] > 0:
            matched_jobs.append({
                "job_id": job.id,
                "title": job.title,
                "company": job.company,
                "location": job.location,
                "job_type": job.job_type,
                "work_mode": job.work_mode,
                "apply_url": job.apply_url,
                **match_data
            })

    return sorted(
        matched_jobs,
        key=lambda job: job["match_percentage"],
        reverse=True
    )

File: artifacts/api-server/src/routes/matches.ts
import { Router, type IRouter, type Request } from "express";
import { eq, or, ilike } from "drizzle-orm";
import {
  db,
  candidateProfilesTable,
  resumeProfilesTable,
  localJobsTable,
} from "@workspace/db";
import { ListMatchesResponse } from "@workspace/api-zod";
import { requireAuth } from "../middlewares/requireAuth.js";
import { calculateMatch } from "../lib/matchingService.js";

type AuthedRequest = Request & { clerkUserId: string };

const router: IRouter = Router();

/**
 * GET /matches — rank local jobs against the authenticated candidate's resume.
 * Filters to location-compatible jobs (preferred location + remote/hybrid).
 * Sorted descending by match percentage. No zero-match jobs returned.
 */
router.get("/matches", requireAuth, async (req, res): Promise<void> => {
  const clerkUserId = (req as AuthedRequest).clerkUserId;

  // Get candidate profile with preferred location
  const [candidate] = await db
    .select()
    .from(candidateProfilesTable)
    .where(eq(candidateProfilesTable.clerkUserId, clerkUserId));

  if (!candidate) {
    res.status(404).json({ error: "Candidate profile not found." });
    return;
  }

  // Get resume profile
  const [resume] = await db
    .select()
    .from(resumeProfilesTable)
    .where(eq(resumeProfilesTable.candidateId, candidate.id));

  if (!resume) {
    res.status(404).json({ error: "Resume profile not found." });
    return;
  }

  // Fetch location-compatible local jobs: preferred location + remote + hybrid
  const preferredLocation = candidate.preferredLocation ?? "Lahore";

  const jobs = await db
    .select()
    .from(localJobsTable)
    .where(
      or(
        ilike(localJobsTable.location, `%${preferredLocation}%`),
        ilike(localJobsTable.workMode, "%remote%"),
        ilike(localJobsTable.workMode, "%hybrid%"),
      ),
    );

  // Calculate match scores
  const matchedJobs: Array<{
    id: string;
    title: string;
    company: string;
    description: string;
    location: string;
    country: string | null;
    jobType: string;
    workMode: string;
    requiredSkills: string[];
    applyUrl: string;
    source: string;
    postedAt: Date | null;
    matchPercentage: number;
    matchedSkills: string[];
    missingSkills: string[];
  }> = [];

  for (const job of jobs) {
    const match = calculateMatch(resume.skills, job.requiredSkills);
    // Skip zero-match jobs
    if (match.matchPercentage === 0) continue;

    matchedJobs.push({
      id: String(job.id),
      title: job.title,
      company: job.company,
      description: job.description,
      location: job.location,
      country: job.country ?? null,
      jobType: job.jobType,
      workMode: job.workMode,
      requiredSkills: job.requiredSkills,
      applyUrl: job.applyUrl,
      source: job.source,
      postedAt: job.postedAt ?? null,
      matchPercentage: match.matchPercentage,
      matchedSkills: match.matchedSkills,
      missingSkills: match.missingSkills,
    });
  }

  // Sort descending by match percentage
  matchedJobs.sort((a, b) => b.matchPercentage - a.matchPercentage);

  res.json(ListMatchesResponse.parse(matchedJobs));
});

export default router;
